src/Kiota.Builder/Writers/TypeScript/CodeConstantWriter.cs

```diff
@@ -21,17 +19,194 @@ public override void WriteCodeElement(CodeConstant codeElement, LanguageWriter w
+        if (codeElement.OriginalCodeElement is not CodeClass codeClass) throw new InvalidOperationException("Original CodeElement cannot be null");
+        if (codeElement.Parent is not CodeFile parentCodeFile || parentCodeFile.FindChildByName<CodeInterface>(codeElement.Name.Replace("NavigationMetadata", string.Empty, StringComparison.Ordinal), false) is not CodeInterface currentInterface)
+            throw new InvalidOperationException("Couldn't find the associated interface for the navigation metadata constant");
+        var navigationMethods = codeClass.Methods
+                                    .Where(static x => x.Kind is CodeMethodKind.IndexerBackwardCompatibility or CodeMethodKind.RequestBuilderWithParameters)
+                                    .OrderBy(static x => x.Name, StringComparer.OrdinalIgnoreCase)
+                                    .ToArray();
+        var navigationProperties = codeClass.Properties
+                    .Where(static x => x.Kind is CodePropertyKind.RequestBuilder)
+                    .OrderBy(static x => x.Name, StringComparer.OrdinalIgnoreCase)
+                    .ToArray();
+        if (navigationProperties.Length == 0 && navigationMethods.Length == 0)
+            return;
+
+        var parentNamespace = codeElement.GetImmediateParentOfType<CodeNamespace>();
+        writer.StartBlock($"export const {codeElement.Name.ToFirstCharacterUpperCase()}: Record<Exclude<keyof {currentInterface.Name.ToFirstCharacterUpperCase()}, KeysToExcludeForNavigationMetadata>, NavigationMetadata> = {{");
+        foreach (var navigationMethod in navigationMethods)
+        {
+            writer.StartBlock($"\"{navigationMethod.Name.ToFirstCharacterLowerCase()}\": {{");
+            var requestBuilderName = navigationMethod.ReturnType.Name.ToFirstCharacterUpperCase();
+            WriteNavigationMetadataEntry(parentNamespace, writer, requestBuilderName, navigationMethod.Parameters.Where(static x => x.Kind is CodeParameterKind.Path or CodeParameterKind.Custom && !string.IsNullOrEmpty(x.SerializationName)).Select(static x => $"\"{x.SerializationName}\"").ToArray());
+            writer.CloseBlock("},");
+        }
+        foreach (var navigationProperty in navigationProperties)
+        {
+            writer.StartBlock($"\"{navigationProperty.Name.ToFirstCharacterLowerCase()}\": {{");
+            var requestBuilderName = navigationProperty.Type.Name.ToFirstCharacterUpperCase();
+            WriteNavigationMetadataEntry(parentNamespace, writer, requestBuilderName);
+            writer.CloseBlock("},");
+        }
+        writer.CloseBlock("};");
+    }
+
+    private static void WriteNavigationMetadataEntry(CodeNamespace parentNamespace, LanguageWriter writer, string requestBuilderName, string[]? pathParameters = null)
+    {
+        if (parentNamespace.FindChildByName<CodeConstant>($"{requestBuilderName}UriTemplate", 3) is CodeConstant uriTemplateConstant && uriTemplateConstant.Kind is CodeConstantKind.UriTemplate)
```

Comment:
Would it make sense to avoid the lookups here by adding an extra property/collection? to the `CodeConstant` so that when creating the constants in the refiner, the `CodeConstantKind.NavigationMetadata` will be annotated with references to the other codeConstants and avoid having to traverse down the namespace  multiple times each time and directly call their references?

Reply:
Because each level is using a dictionary lookup under the covers, because we're limiting the search to 3 levels deep, and because we're doing this in the writer, which runs un parallels for all the elements, the performance impact should be limited.
I thought about adding short hands in the DOM elements, but the challenge is that is creates a tree over the tree, which might be really hard to keep in sync in case of structure changes.
What I'm more worried about is the naming conventions changing for whatever reason (like a PM changing their minds 😋) and replacements not being done consistently leading to broken things. We should move that as constants probably. Thoughts?

Reply:
Yeah. Now that the depth limited search this is clearer to me. I'm happy to leave this as is. 

> We should move that as constants probably. Thoughts?

I agree. We should use the constants so that the naming and the searching are easily more consistent.

Reply:
added the constants, resolving